build-utils.js

```diff
@@ -94,41 +97,85 @@ function taskHarness(task, projectOrStar, ...args) {
+    ];
 
-  if (options.outputName.indexOf('build/') !== 0) {
-    throw new Error('Expected options.ouputName to start with \'build/\'');
-  }
+    gulp.src(sources)
+      .pipe(sourcemaps.init())
+      .pipe(rollup(options.rollupConfig))
+      .pipe(gulpif(options.minify, babel({
+        presets: ['babili', {comments: false}],
+      })))
+      .pipe(header(LICENSE_HEADER))
+      .pipe(rename(outputName))
+      .pipe(sourcemaps.write('.'))
+      .pipe(gulp.dest(`${options.projectDir}/build`))
+      .on('error', reject)
+      .on('end', resolve);
+  });
+}
 
-  return new Promise((resolve, reject) => {
-    gulp.src([
-      path.join(options.projectDir, 'src', '**', '*.js'),
-      path.join('lib', '**', '*.js'),
-      path.join('packages', '**', '*.js'),
-      path.join('node_modules', '*', '**', '*.js'),
-    ])
-    .pipe(sourcemaps.init())
-    .pipe(rollup(options.rollupConfig))
-    .pipe(babel({
-      presets: ['babili', {comments: false}],
-    }))
-    .pipe(header(licenseHeader))
-    .pipe(rename(options.outputName.substring('build/'.length)))
-    // Source maps are written relative tot he gulp.dest() path
-    .pipe(sourcemaps.write('.'))
-    .pipe(gulp.dest(destPath))
-    .on('error', (err) => {
-      reject(err);
-    })
-    .on('end', () => {
-      resolve();
+/**
+ * A helper to generate Rollup build configurations.
+ *
+ * One build config is generated for each format given in formatToPath.
+ * Both minified and unminified build configs are generated.
+ *
+ * @param {Object.<String,String>} formatToPath A mapping of each format
+ *        ('umd', 'es', etc.) to the path to use for the output.
+ * @param {String} projectDir The path of the project directory.
+ * @param {String} moduleName The name of the module, used in the UMD output.
+ * @param {Array.<Object>} [plugins] A list of Rollup plugins to use.
+ * @returns {Array.<Object>}
+ */
+function generateBuildConfigs(formatToPath, projectDir, moduleName, plugins) {
+  // This is shared throughout the full permutation of build configs.
+  const baseConfig = {
+    rollupConfig: {
+      entry: path.join(projectDir, 'src', 'index.js'),
+      plugins,
+      moduleName,
+    },
+    projectDir,
+  };
+
+  // Use Object.assign() throughout to create copies of the config objects.
+  return [true, false].map((minify) => {
+    return Object.assign({}, baseConfig, {minify});
+  }).map((partialConfig) => {
+    return Object.keys(formatToPath).map((format) => {
+      const fullConfig = Object.assign({}, partialConfig, {
+        buildPath: formatToPath[format],
+      });
+      fullConfig.rollupConfig = Object.assign({}, partialConfig.rollupConfig, {
+        format,
+      });
+
+      // Remove the '.min.' from the file name if the output isn't minified.
+      if (!fullConfig.minify) {
+        fullConfig.buildPath = fullConfig.buildPath.replace('.min.', '.');
+      }
+
+      return fullConfig;
     });
-  });
+  }).reduce((previous, current) => {
```

Comment:
This is probably due to the chaining of maps and reduce. But what is the input and output of this? I thought I was following this but apparently not.